### DISCUSSION TITLE
End of support for Vite/Svelte and Astro.

### DISCUSSION THREAD

**Discussion by endigo9740** [skeletonlabs/skeleton#721] [Announcements]
Hey everyone, I wanted to take a moment to provide a detailed explanation for this change. We hope this sheds some light on our reasons and decision not to move forward with these frameworks. Please be aware the goal will be to put our full focus on SvelteKit going forward, as we feel this provides the best experience for users going forward.

## The Impact on Components

When Skeleton was open sourced in August the goal was to support it everywhere Svelte was available. We were almost immediately alerted to the fact Skeleton wasn't working as expected for Vite/Svelte and Astro projects. We immediately opted to change several components and features to facilitate these frameworks.

One example was our [LightSwitch](https://www.skeleton.dev/utilities/lightswitches) component, which is used toggle the preferred light/dark mode for themes. It wasn't supported in these frameworks due to the fact were using a SvelteKit-only `browser` environment context to conditionally check when LocalStorage became available. Because of this we opted for Svelte's `onMount` lifecycle hook. While it made the feature "universal", it came with a noticeable impact on speed. In fact, you may may still notice a quick "flash" on page load due to this - which we aim to address in the near future.

## Reviewing Audience Usage

As the Skeleton community continues to grown, we've been able to poll you (our audience) and get a rough estimate of the usage per each framework. Here's the results from around a month ago:

![image](https://user-images.githubusercontent.com/1509726/208256901-b0c96417-ad2d-4ce3-b23d-6b8d86acd5ac.png)

![Screen Shot 2022-12-17 at 12 45 51 PM](https://user-images.githubusercontent.com/1509726/208256904-ffe89211-80bd-4a1f-9a44-0cc189bab90f.png)

While Vite/Svelte and Astro are represented here, you'll note the numbers are VERY small. Including 0% for Astro on Twitter.

## The Launch of SvelteKit v1.0

You may also be aware that [SvelteKit v1.0 launched this week](https://svelte.dev/blog/announcing-sveltekit-1.0). It's been noted in the community that Svelte seems to be taking a step away from Vite/Svelte. If you view the Svelte homepage you'll see the CLI install commands now use the same "create svelte" tool which creates a SvelteKit app:

![Screen Shot 2022-12-17 at 12 54 04 PM](https://user-images.githubusercontent.com/1509726/208257231-060213eb-b8be-46ca-a8ab-c7090aeb25d7.png)

![Screen Shot 2022-12-17 at 1 00 18 PM](https://user-images.githubusercontent.com/1509726/208257504-b3194860-98a9-495a-8aee-85dada98b8fe.png)

From what we're hearing, the goal is to make SvelteKit the "de facto" option, even for folks using Svelte in embedded solutions like browser extensions or web-based desktop stacks like Tauri and Electron. While out of our control, we are hopeful this becomes the case.

## The Cost of Support

If you're not aware, we actually maintain unique onboarding instructions for Vite and Astro users, dedicated tutorials, and cross test per each major feature or change. We also provide free and voluntary support between GitHub and Discord for these frameworks. But there's no getting around it, the core team of Skeleton maintainers is small. Primarily made up of myself, @niktek, @Sarenor, and @luke-hagar-sp. With my recent appearance on Svelte Sirens we've seen a big influx of new users and an even larger number of new support tickets. While we love helping everyone, it does come at the cost of our development time. Therefore a balance must be struck.

## Removing Support

Giving the information above, we've made the call to drop support for Vite/Svelte and Astro in the coming release. This is not a decision we've reached lightly, as you can see we've been polling and reviewing information for weeks now. But honestly the numbers just aren't there to justify continued support. That said, it's worth noting that a portion of Skeleton features, such as Tailwind Elements, will continue to work fine for these frameworks. If you're looking purely for style and theme solutions then you most likely can continue to use these without issue.

## What Comes Next

The first order of business will be to [update and simplify the onboarding documentation](https://github.com/skeletonlabs/skeleton/pull/723) and go all-in on SvelteKit and our CLI. We feel this provides the best solution for most. We can continue to improve the CLI while also starting the process of updating features to support optimal solutions for the majority of our users.

If you have any questions or feedback regarding this change, please reach out in the comment section below this post. We would be glad to hear from you.

Thanks for your time!
Chris

**Comment by wiverson** [comment]
Just as an FYI, I'm using SvelteKit w/Capacitor in SPA only mode and it seems to be working fine (more or less). I've also used Astro and Tailwind to build a blog and TBH the Svelte usage there is really just to add a bit of pop - as soon as it turns into SPA-like routing thing I'd probably start with SvelteKit as a base instead.

+1

**Reply by endigo9740** [reply]
Thanks for sharing @wiverson. That matches the experiencing I hear from other users. We know this is going to affect a few folks, and we hate that, but hopefully the post makes our reasons clear for the change in direction. I really do hope that SvelteKit continues to grow and evolve to the point it satisfies all use cases. It's just a great piece of tech IMO!

**Reply by risalfajar** [reply]
Is using Svelte Kit with SPA mode the same with using Svelte?
I'm considering to migrate after reading this announcement.

**Reply by endigo9740** [reply]
Not exactly. If you're familiar with other libraries, the easiest way to explain it is to compare it to it's peers:

- Svelte is UI component library comparable to React and Vue
- SvelteKit is an app framework comparable to Next.js and Nuxt

SvelteKit brings lots of goodies to the table:

- Routing
- SSR
- Special environments (ex: `browser`)
- Adapters - which enable building for SPA, SSR, etc.
- Plus lots of little extras, such as preloading via hover on links, etc

Vite + Svelte stops being practical the moment you get past a rudimentary one-page app IMO. Having to install your own router is not always a trivial task. Even if you just use SvelteKit to write Svelte components + routing, it's exceptionally good in that role.

**Reply by risalfajar** [reply]
Sorry, my question isn't clear enough.

What I mean is, can I build an SPA with SvelteKit then upload it into static web hosting (Firebase Hosting)? I don't need SSR.

**Reply by Sarenor** [reply]
That should be about what adapter-static does, as far as I understood it, yeah: https://github.com/sveltejs/kit/tree/master/packages/adapter-static#spa-mode

**Reply by endigo9740** [reply]
Oh sorry, yeah absolutely. We hosted the Skeleton doc site on S3 in that manner for most of last year. Works fine!

**Comment by noname-oni** [comment]
Omg, I saw a push in the morning and at first I thought that there would be no svelte support at all. Wow, that was very scary.

**Reply by endigo9740** [reply]
I can understand the mistake. We supported three "frameworks" previously:

1. SvelteKit - this will remain
2. Vite + Svelte - support for this will end
3. Astro - support for this will end

In a sense we're going all in on SvelteKit, just like Svelte themselves, which a large majority of our audience uses. We have no plans to stop supporting Svelte though, we love it!

**Comment by Scarecro** [comment]
Hi. I'm willing to use the old versions, or deal with the hassle of losing support, because I would rather have Skeleton in my project than not. It's the best fit of components for my Astro app.
 
 Is there anywhere I can find help or instructions for installing it? (V2 preffered)

**Reply by endigo9740** [reply]
Hey Scarecro, we are no longer provide SUPPORT for Astro, but Skeleton still works within Astro projects. We've actually gone out of our way to preserve support for Vite or Astro projects, we just don't maintain documentation regarding this or help troubleshoot support issues for this specifically.

To get started, I'd recommend just reading through the Getting Started instruction for v2 (which is a release candidate now, and stable dropping in the next day or so), which should be simpler than v1. Less Sveltekit-specific setup required:
https://skeleton-docs-git-v2-skeleton-labs.vercel.app/docs/get-started

You'll need to translate some of the instruction, but overall it should be simple enough to translate. Layouts being different in Astro obviously. I believe Astro also has it's own way to implement Tailwind as well.

**Reply by Scarecro** [reply]
After lots of trial and error I can see the themes active on Skeleton components, and tailwind works fine. Your themes are amazing btw!

Most things on the Svelte side don't work though. Is it a case where something used to be in the instructions to handle SSR that I'm haven't figured out? 

I checked waybackmachine, but I wasn't able to press the Astro tab to see what was written there last year. Is there any place that I can find those old instructions? They could probably provide useful hints.

**Reply by endigo9740** [reply]
We did have some documentation around this. Off hand it involved wrapping Svelte components in Astro components to ensure props "outside" the Svelte component would work as expected, then using a particular hydration method on the Astro version. It was all a bit complicated and convoluted. I'm not entirely sure it was the right way to go about it, but I was never an expert at Astro either.

If you look in the Skeleton project history the docs for this should still be present, just note this was when the Docs + Core library were a single project, not the monorepo structure we use now. I'd look for a version of the repo from around December last year for this.

**Reply by Scarecro** [reply]
Wow. You're right, that's definitely not a good great way to do things. The fact that it worked though is amazing, and I now know where to go to continue my research. 

In the meanwhile I'm trying to branch my app and see how much code will work on SvelteKit. Thankfully I have more Svelte components than react or Astro ones. I'd rather rewrite components and server code now so I can use Skeleton, than need to manually restyle the entire app later.

Thx for the help. It's comforting to still have solid options if porting it over proves to be unfeasible.
